File: basic/bruteforce/mazesearcher/MazeSearcher.java
// 문제 링크 : 
// https://www.acmicpc.net/problem/2178

// [solution] : bfs를 이용
// 미로를 2차원 배열 maze에 저장하고,
// Point 클래스를 이용해 
// 현재 위치 (r, c)를 큐에 넣어 BFS를 수행
// 이동 가능한 칸을 방문할 때, 
// 현재 칸까지의 거리 값에 1을 더해 maze에 저장, visited 배열에 체크 표시


import java.io.*;
import java.util.*;

public class MazeSearcher {
    static int n, m;
    static int[][] maze;
    static boolean[][] visited;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(
            new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(
            br.readLine());
        
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        maze = new int[n][m];
        visited = new boolean[n][m];
        
        for (int r = 0; r < n; r++) {
            String line = br.readLine().trim();
            for (int c = 0; c < m; c++) {
                maze[r][c] = line.charAt(c) - '0';
            }
        }

        bfs(0, 0);
    
        System.out.println(maze[n - 1][m - 1]);
    }

    static class Point {
        int r, c;
        public Point(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    static void bfs(int sr, int sc) {
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(sr, sc));
        visited[sr][sc] = true;
        
        while (!queue.isEmpty()) {
            Point cur = queue.poll();
            int cr = cur.r;
            int cc = cur.c;

            for (int i = 0; i < 4; i++) {
                int nr = cr + dr[i];
                int nc = cc + dc[i];

                if (nr < 0 || nc < 0 || nr >= n 
                || nc >= m) {
                    continue;
                }

                if (maze[nr][nc] == 1 
                && !visited[nr][nc]) {
                    maze[nr][nc] = maze[cr][cc] + 1;
                    visited[nr][nc] = true;
                    queue.offer(new Point(nr, nc));
                }
            }
        }
    }
}

File: basic/bruteforce/mazesearcher/MazeSearcher.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <string>
using namespace std;

int bfs(vector<vector<int>>& maze, int n, int m) {
    // 상하좌우 이동을 위한 방향 벡터
    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    queue<pair<int, int>> q;
    q.push({0, 0});

    while (!q.empty()) {
        auto [r, c] = q.front();
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            // 미로 범위 체크 및 이동 가능한 칸 (1)인지 확인
            if (nr >= 0 && nr < n && nc >= 0 && nc < m && maze[nr][nc] == 1) {
                maze[nr][nc] = maze[r][c] + 1;
                q.push({nr, nc});
            }
        }
    }
    
    return maze[n - 1][m - 1];
}

int main() {
    int n, m;
    cin >> n >> m;
    // 개행 문자 제거를 위해 cin.ignore() 사용
    cin.ignore();

    vector<vector<int>> maze(n, vector<int>(m));
    string line;
    for (int i = 0; i < n; i++) {
        getline(cin, line);
        for (int j = 0; j < m; j++) {
            // 문자 하나씩 정수로 변환 (문자 '0' 또는 '1'을 정수로 변환)
            maze[i][j] = line[j] - '0';
        }
    }
    
    cout << bfs(maze, n, m) << endl;
    return 0;
}
